keps/sig-storage/20190117-SkipVolumeOwnershipChange.md

```diff
@@ -0,0 +1,147 @@
+
+### Implementation Details/Notes/Constraints [optional]
+
+When creating a pod, we propose that `PersistentVolumeClaimVolumeSource` field expanded to include a new field called `PermissionChangePolicy` which can have following possible values:
+
+ - `NoChange` --> Don't change permissions and ownership.
+ - `Always` --> Always change the permissions and ownership to match fsGroup. This is the current behavior and it will be the default one when this proposal is implemented.
+ - `OnDemand` --> Only change permissions when fsGroup of pod and PVC don't match.
+
+```go
+type PersistentVolumeClaimVolumeSource struct {
+    // ClaimName is the name of a PersistentVolumeClaim in the same namespace as the pod using this volume
+    ClaimName string
+    // Optional: Defaults to false (read/write).  ReadOnly here
+    // will force the ReadOnly setting in VolumeMounts
+    // +optional
+    ReadOnly bool
+    // PermissionChangePolicy ← new field
+    PermissionChangePolicy string
+}
+```
+
+In addition to this, we propose a new field to PVC's status:
+
+```go
+type PersistentVolumeClaimStatus struct {
+     // Phase represents the current phase of PersistentVolumeClaim
+     // +optional
+     Phase PersistentVolumeClaimPhase
+     // AccessModes contains all ways the volume backing the PVC can be mounted
+     // +optional
+     AccessModes []PersistentVolumeAccessMode
+     // Represents the actual resources of the underlying volume
+     // +optional
+     Capacity ResourceList
+     // FSGroup of PVC
+     // + optional
+     FSGroup *int64     // <------ NEW ------
+     // +optional
+     Conditions []PersistentVolumeClaimCondition
+}
+```
+
+
+When a volume is mounted by the kubelet, volumemanager will check `pvc.Status.FSGroup` and `pod.SecurityContext.FSGroup` and depending on 
+`PermissionChangePolicy` of `PersistentVolumeClaimVolumeSource` it will:
+
+ - Do no permission and ownership change if `PermissionChangePolicy` is `NoChange`.
+ - Will change ownership and permission of volume to match `pod.SecurityContext.FSGroup` if `PermissionChangePolicy` is set to `Always`.
+ - if `PermissionChangePolicy` is set to `OnDemand`, volume permissions will ONLY be changed if `pvc.Status.FSGroup` and `pod.SecurityContext.FSGroup` don't have the same value.
+
+After permissions are changed, pvc.Status.FSGroup will be updated to reflect latest value.
+
+### Risks and Mitigations
+
+## Graduation Criteria
+
+How will we know that this has succeeded?
+Gathering user feedback is crucial for building high quality experiences and SIGs have the important responsibility of setting milestones for stability and completeness.
+Hopefully the content previously contained in [umbrella issues][] will be tracked in the `Graduation Criteria` section.
+
+[umbrella issues]: https://github.com/kubernetes/kubernetes/issues/42752
```

Comment:
Is this the right link? It's pointing a closed issue?

It would be great to have a link to a test plan with test cases, and add tests passing to the graduation criteria.

Reply:
we will update this link to - https://github.com/kubernetes/kubernetes/issues/69699 , does that work?

Reply:
@msau42 Could you provide example KEPs (or an example PR) with the kind of monitoring you're looking for?  I'm not able to find any that reference monitoring in a way that I can find applicable to this feature.  I'm not fully aware of all the monitoring practices within Kubernetes, so I do not know what is appropriate to add here.

Reply:
We can skip monitoring for now IMO. It is important for stuff that emit metrics etc. I do not think we need to worry about it now.

Reply:
I think as part of the alpha->beta criteria, we should add in some sort of metric.

For example, the issue that this is trying to solve is that mount can take a long time if you have a large volume.  Do we have metrics in place that measure this recursive chown time, so that we can tell if this feature is effective in that regard?

Reply:
Okay, it could be nice-to-have not must-have. Because whenever I have debugged such issues, since we already have metric for mount timings it is relatively easy to figure out, if it is mount that is taking longer time or `chown/chmod`. Do you think this is a blocker for merging @msau42 ?

Reply:
I think we should have a distinguishing metric for the actual mount call vs fsgroup ownership change.  I have seen issues where the mount call hangs when the disk has IO/network issues and our existing metric cannot distinguish between the two.

Reply:
I do not disagree but I don't know why this KEP must be the one that should be implementing it.  The metric is probably useful in general and should exist regardless of this KEP gets implemented or not.

Reply:
I think the need for a separate metric becomes more important with this feature.  Because we cannot reliably tell if this feature is working with the current metric

Reply:
The metric has been added to the spec and to the Beta graduation criteria